### DISCUSSION TITLE
Which response type should wrapException have?

### DISCUSSION THREAD

**Discussion by sergioflores-j** [sergioflores-j/ts-to-go#67] [Ideas]
## Current scenario

Currently we return a tuple of two positions:

```
[error: unknown, response: ReturnType<wrappedFunction> | undefined]
```

This has some limitations, as also described here:
#8 and #29 

In short, TS is unable to determine if the response the consumer got was an error or a success in a easy/well typed way. 
That is because `unknown` is the strongest type and removes the `undefined` definition.

```ts
const myWrappedFn = wrapException(async () => 'string');
const [error, response] = await myWrappedFn();

// error: unknown ; response: string | undefined
if (error) {
  // do error handling
  return;
}

// The type here is still:
// error: unknown ; response: string | undefined
// whereas it should be:
// error: undefined ; response: string
```

This has forced consumers to create utility functions like the one here: #8 

### Tuple with extra parameter

You can use this Typescript playground to check: [TS Playground tuple](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBDAnmYcCiUrQErAM6QB2ewAPBllACrL5wC8cAroQNaEQDuhAfA3FGABDACYRCAG0RwA2gCg4iuMEzQAXOlXVaeADQKlgguJIaWI4ADMAloWAj9SuNbwV1CKE2D6AugG45OVBIWARaOABlJgBjaPw8XGNiMkSiEn4Wdi5efkFRcSlZA0UVSjNCCxs7B2KBfDTgDVSTb1qXNygNSyEJEl8AoPBoeCQUOAB1KCEwFBFm5NJ5vs1KGhQ8PkYOpbIOtfw+AB9ImLi8BPqWxcvkngHg4bCxyenZ0gMAMUJlEBhgCrwcAAFAA6MFCKAAczw5Sy3BkPgAlAw+JkONxHCtoPtAYw0dlMTt+LgYEwoIR9qQvjx9JtgQYwSCIdCNAAFCFCAC2wD+UDwVN4cmR9D4RNAfwBcFZWE5LmAABlrKwyLZLCopTK5SSyRTaDwDAB+DUQWUkUgvGb2HakaUmrU8nX7XRY7TrHj6xQaC2za07Z17HR3QIWaISCGoSwsaIwaziOCcKZgNAgOJgGPiAU-CUiQGg8FQmFwISERAI4V8YuIf1aHE8IGWQgaL6Ir2Jt5faurQOBAD0PbgAFoh8OR6PR3I+3AqGgIlQ4ABhACCERnE-7Y43G7k0RM8E5iC+-CEeEQhGiwOZGjwMCgtkhLeNprI19vhEhCM2fAA3m1LBeociMAABZYJwcB2GBHRAgA5AARmG0GIrUN6IF+gikuSRacEI1jwLaT4gkYEASAAbsAQIyNBL53tBSJ+AAvnA0RCDA0RAUCKjIl+wGgcoUD0XIAk7sQe6IBEp7RIejBApecDUW+5ZwD+TjWH+MkAQgIFcOBwCQVoMHwUIiG1OhOpwFRN40QE9EDJOaAABqLgAsqy8poHAACqESLgA4mga5wAAEgA8gAamg2BwGFEVToF7mhYu8oeTOU7BQucXzgA0rFaAAJKRVQACarKrsJ15wA2R4nmewKKcpShlfAx4SV83r2PwCbTMmqbpoQQL7s2AROLUk7LoVABy861I1silNAzpGA0zrtFoPhHthuFFtVkmEG1IhAgADIiQ1KL+wIrZQXG1E4ZXEcAIISBAkIcVoSFODdJh3Q9T1AotLRve9iimeSJ2KAJw0fcQX2Pc9c1QADii3RI90w79NwkCCnLTDJKJFoiCNwCN-YRBNU2Q+V22tW27WMJ1SYpsAaaxn1+7iWeg3Tbus1aGz0QLejwC88trg8xJa2MJTu3U-tR0DCpakXdAvNXYDSMoz9cPK6DDWfcj33PX9yRa9dhgOiDtQCZzUN66jmsScdVt4NDP2GyQxv0UAA)

The definition would be something like:
```ts
export type ErrorResponse<ErrorTypes = any> = readonly [
    error: ErrorTypes,
    response: undefined,
    isError: true,
];
export type SuccessResponse<Response = any> = readonly [
    error: undefined,
    response: Response,
    isError: false,
];
```

Consuming this would be something like:

```ts
const [error, response, isError] = await myWrappedFn();

// error: unknown ; response: string | undefined; isError: boolean

if (isError) {
  // error: unknown; response: undefined
  console.error(error);
  // do error handling
  return;
}

// The type here now is:
// error: undefined; response: string
```

### Pros

- This is probably not a breaking change and consumers can still use the functions the way they did before with the addition of this new property.
- With the tuple the error is the most important part, and forces the consumer to think about it.

### Cons
- It appears to be quite troublesome to set variables in files using multiple wrapped functions

```ts
const [error, response, isError] = await myFn1();

// code gets creative here
const [errorMyFn2, responseMyFn2, isErrorMyFn2] = await myFn2();
const [errorMyFn3, responseMyFn3, isErrorMyFn3] = await myFn3();
```

### Object response

You can use this Typescript playground to check: [TS Playground object response](https://www.typescriptlang.org/play/?#code/FAUwHgDg9gTgLgAjgTwiBBRGNYCUQDO0AdgSADxY4wAqqhCAvAgK7EDWxUA7sQHxMEAb2AIEASwJVYALiQwWIANyiEIbLMwba9AirEB6AFRHVCIwgAC3AIYxi44gHMEAYwAWIV+wQADSdIwvggQOGjwyAgARiAAZrDoNq6uhASOLr4AJjZwNsGwfurUvmZGBqrZuXJsnDzEKgC+KqCQsIgoaAgAyizJqfhEUKQUAyRkgjVcvALMImIB2nKxNgA2ZPpqi6zEmXGOIJkblTZyo0PrwE3ALdDwSPQIAOowNhBomWfD5J9kADRa1DoaAIMwBeEIYwogSBhAEAB9ur0UgQCD8RhDziA+M1wLd2g9nq93uRVAAxYhqMBwEA7AgIAAUADpmXYnARqhwpsQANoAXQAlEwBJM6r9VNDdBNOaLVGjBPg4Cx7DDyOS+GLQfTVMzGaz2QgAAp2GwAWxA1JgBFV-GAgsYAjl4GptMNOBNkhAABlxOwKI5YupXVB3WQFUriDC+KoAPxBkMUQlvA5o8gGt0esPK+j-CXAvhRsRyRPvFNonPaGEg5q7VwrOzoWJsVxwcRDBDcF4QDBgFIQFtDa2U52ZOlMlkwNlyGzEZB8u0CafIcuA3R8emxYhycn8oud4nk5ewStR4AGAwIAC0V+vN9vt9P55oGC6NAQAGEAIJdZ8Py93-93sArjnIgJrIOSgg2AQyDEK4DKsnIBBwDA6Q7nGHrkEhKHOHyMwCHMYgSLE8EToKcDuDg3AIMQIBUYE9IAORRHWDH8mY8jIEIMDmuGCA2LY4iIGmwYeoy3EEFAKwAG4gPS3IMVh6QMQKSgNG4OQePS6iCkI5GUZsDSXEBIEIGBXQwa4EHMPSCEIIpzjzsI7HiMRNmkUgFE8NRtFgjAjHMTYrHsdxir2AgCnIUpjTNGemAABofgAsganoYAgACqXQfgA4hgv4ABIAPIAGoYLgCAlWVCA0PlaXFR+nrpc+1WFe+tVvgA0tVtUAJLlTQACaBo-sBpCIBukHQbBDKOQRYijUhfFTZZxDFgcggdq83a9v2xD0mB24bGI7GxV+A0AHJvuxC2IEIRSwP8xz-As1Bqcw-E2IJS0WeSa2ZPSAAM-JHQgzmuS9sA6exhFuOckkgIyKxQE4WnaGxMMwwt8OI8j9LHOjGNiCF4Yg4Z11wysCNIyj90wAT80U1TuPHIyJqvDZQp8fy9OgzDsVdBdV2YyZy2-Xu63MJtXY9iAfatntZk-cQwPk2NdkWXKzCi6t4v-UDzQwy5DLLWijIQ3TTmE1jlM4yjJsYsMjK0yrVuM7b9L24Mjv4yDMPE-YpPXMLpDY9THsaw7ZBO2jIPW0zdsR17Uc+5cQA)

The definition would be something like:

```ts
export type ErrorResponse<ErrorTypes = unknown> = {
  isError: true;
  error: ErrorTypes;
  /**
   * @warning check `isError` property before accessing `data` or `error`
   */
  data: unknown;
};

export type SuccessResponse<Response = unknown> = {
  isError: false;
  error: undefined;
  data: Response;
};
```

Consuming this would be something like:

```ts
const myFnResponse = await myWrappedFn();

// myFnResponse = { error: unknown ; data: unknown; isError: boolean }

if (myFnResponse.isError) {
  // myFnResponse = { error: unknown ; data: unknown; }
  console.error(myFnResponse.error);

  // do error handling
  return;
}

// The type here now is:
// myFnResponse = error: undefined; data: string
```

### Pros

- With objects the code can grow indefinitely as obtaining the results for wrapped functions can be groupped in a single variable

```ts
const myFn1Response = await myFn1();

const myFn2Response = await myFn2();
const myFn3Response = await myFn3();
```

- We can enforce the consumer to check the `isError` by setting the `data` to unknown otherwise.

### Cons

- It is a breaking change for the library
- Error attribute can be easily "ignored", as now the most important attribute becomes `isError`.
- It is slightly harder to destructure and rename the values from the response object.

### To be defined:

Attribute names 
```ts
isError
data
error
```

or 

```
success
data
error
```

Which is best? Other options?

**Comment by MarcioMeier** [comment]
I'd rather to stick with the object approach because with the tuple approach makes hard to follow a flow when it comes with 2+ wrappers, as the names are likely to be very similar.

Also, Javascript engineers are quite used to work with objects so it shouldn't be hard to follow the new interface.

Regarding the attributes names, I'd go with `isError`, as this library mostly is about handling error, it makes sense to highlight it.

I'm just not sure about the `is`Error because the result is not an error itself. Maybe `hasError` is more intuitive

**Reply by flaviostutz** [reply]
I like the object approach also and agree with the arguments that JS devs are used to handle this kind of returns whereas returning destructured array is not that common at all... it also makes things much better typed.

**Comment by sergioflores-j** [comment]
Then object it is!
Thanks for the input!

Implemented here #30 
And will be available in v2.0.0-beta.1 for initial test rollout.
